eventuate-tram-sagas-reactive-common/src/main/java/io/eventuate/tram/sagas/reactive/common/ReactiveSagaLockManagerImpl.java

```diff
@@ -0,0 +1,108 @@
+package io.eventuate.tram.sagas.reactive.common;
+
+import io.eventuate.common.jdbc.EventuateDuplicateKeyException;
+import io.eventuate.common.jdbc.EventuateSchema;
+import io.eventuate.common.json.mapper.JSonMapper;
+import io.eventuate.common.reactive.jdbc.EventuateReactiveJdbcStatementExecutor;
+import io.eventuate.tram.messaging.common.Message;
+import io.eventuate.tram.messaging.producer.MessageBuilder;
+import io.eventuate.tram.sagas.common.SagaLockManagerSql;
+import io.eventuate.tram.sagas.common.StashedMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.Map;
+
+public class ReactiveSagaLockManagerImpl implements ReactiveSagaLockManager {
+
+  private Logger logger = LoggerFactory.getLogger(getClass());
+
+  private EventuateReactiveJdbcStatementExecutor eventuateJdbcStatementExecutor;
+  private SagaLockManagerSql sagaLockManagerSql;
+
+  public ReactiveSagaLockManagerImpl(EventuateReactiveJdbcStatementExecutor eventuateJdbcStatementExecutor,
+                                     EventuateSchema eventuateSchema) {
+
+    this.eventuateJdbcStatementExecutor = eventuateJdbcStatementExecutor;
+
+    sagaLockManagerSql = new SagaLockManagerSql(eventuateSchema);
+  }
+
+  @Override
+  public Mono<Boolean> claimLock(String sagaType, String sagaId, String target) {
+    return eventuateJdbcStatementExecutor
+            .update(sagaLockManagerSql.getInsertIntoSagaLockTableSql(), target, sagaType, sagaId)
+            .then(Mono.just(true))
+            .onErrorResume(EventuateDuplicateKeyException.class, e -> {
+              Mono<String> owningSagaId = selectForUpdate(target);
+              return owningSagaId.map(id -> id.equals(sagaId)).switchIfEmpty(claimLock(sagaType, sagaId, target));
+            });
+  }
+
+  private Mono<String> selectForUpdate(String target) {
+    return eventuateJdbcStatementExecutor
+            .query(sagaLockManagerSql.getSelectFromSagaLockTableSql(), target)
+            .map(row -> row.get("saga_id").toString())
+            .collectList()
```

Comment:
I believe that this is https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#next--

Reply:
I will read the docs